llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -420,11 +420,31 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
   return NF;
 }
 
+/// Returns true if the Ptr is loaded by any Load in the vector of
+/// Loads, and if the Loaded value is not a pointer.
+static bool checkIfPointerIsDereferenced(SmallVector<LoadInst *, 16> &Loads,
```

Comment:
With the most recent changes, please check again if checkIfPointerIsDereferenced(); I suspect the other checks make this unnecessary.

Reply:
This is still needed to prevent the failure in (pr42028-recursion.ll) case, as it is not self recursive. But, if I add the self recursive check I mentioned above, then it would be okay.